wandb/sdk/launch/agent2/controllers/k8s.py

```diff
@@ -64,90 +57,94 @@ def __init__(
+                owned_items[next_item.id] = next_item
 
         # ensure all our owned items are running
-        for item in owned_items:
-            if item["id"] not in self.active_runs:
+        for owned_item_id in owned_items:
+            if owned_item_id not in self.active_runs:
                 # we own this item but it's not running
-                await self.launch_item(item)
+                await self.launch_item(owned_items[owned_item_id])
 
+        # TODO: validate job set clears finished runs
         # release any items that are no longer in owned items
         for item in self.active_runs:
             if item not in owned_items:
                 # we don't own this item anymore
                 await self.release_item(item)
 
-    async def lease_next_item(self) -> Any:
-        raw_items = list(self.jobset.jobs.values())
-        pending_items = [item for item in raw_items if item["state"] in ["PENDING"]]
-        if len(pending_items) == 0:
-            return None
-
-        sorted_items = sorted(
-            pending_items, key=lambda item: (item["priority"], item["createdAt"])
-        )
-        next_item = sorted_items[0]
-        self.logger.info(f"Next item: {json.dumps(next_item, indent=2)}")
-        lease_result = await self.jobset.lease_job(next_item["id"])
-        self.logger.info(f"Leased item: {json.dumps(lease_result, indent=2)}")
+    async def pop_next_item(self) -> Optional[Job]:
+        next_item = await self.queue_driver.pop_from_run_queue()
+        self.logger.info(f"Leased item: {json.dumps(next_item, indent=2)}")
+        return next_item
 
-    async def launch_item(self, item: Any) -> Any:
+    async def launch_item(self, item: Job) -> Optional[str]:
         run_id = await self.launch_item_task(item)
         self.logger.info(f"Launched item got run_id: {run_id}")
         return run_id
 
-    async def launch_item_task(self, item: Any) -> str:
+    async def launch_item_task(self, item: Job) -> Optional[str]:
         self.logger.info(f"Launching item: {json.dumps(item, indent=2)}")
 
-        project = LaunchProject.from_spec(item["runSpec"], self.legacy.api)
+        project = LaunchProject.from_spec(item.run_spec, self.legacy.api)
         project.queue_name = self.config["jobset_spec"].name
         project.queue_entity = self.config["jobset_spec"].entity_name
-        project.run_queue_item_id = item["id"]
+        project.run_queue_item_id = item.id
         project.fetch_and_validate_project()
         run_id = project.run_id
         job_tracker = self.legacy.job_tracker_factory(run_id)
         job_tracker.update_run_info(project)
 
-        assert (
-            project.docker_image is not None
-        ), "Docker image must be set when using k8s"
-        run = await self.legacy.runner.run(project, project.docker_image)
+        ack_result = await self.jobset.ack_job(item.id, run_id)
+        self.logger.info(f"Acked item: {json.dumps(ack_result, indent=2)}")
```

Comment:
nit: Move this below the `if not ack_result:` block

Reply:
(The GitHub UI doesn't make it obvious but I'm only referring to the `self.logger.info` line here)

Reply:
nice catch